File: examples/typescript/express-server/src/client-example.ts
/**
 * Example X402 Client Application
 *
 * Demonstrates how to use the X402AutoClient to access paid API endpoints.
 */

import { Keypair } from "@solana/web3.js";
import { X402AutoClient, X402Client } from "@openlibx402/client";
import * as fs from "fs";
import * as path from "path";

// Example 1: Using X402AutoClient (Automatic Payment Handling)
async function exampleAutoClient() {
  console.log("\n=== Example 1: Auto Client (Automatic Payments) ===\n");

  // Load wallet keypair (in production, load from secure storage)
  // For this example, you would need to create a keypair
  // const keypair = Keypair.fromSecretKey(
  //   Uint8Array.from(JSON.parse(fs.readFileSync('wallet.json', 'utf-8')))
  // );

  // For demo purposes, generate a random keypair
  const keypair = Keypair.generate();
  console.log("Wallet Address:", keypair.publicKey.toString());

  // Create auto client
  const client = new X402AutoClient(keypair, "https://api.devnet.solana.com", {
    maxPaymentAmount: "5.0", // Safety limit
  });

  try {
    // Access free endpoint (no payment required)
    console.log("Fetching free data...");
    const freeResponse = await client.get("http://localhost:3000/free-data");
    console.log("Free data:", freeResponse.data);

    // Access premium endpoint (payment required)
    // Note: This will fail without real funds in the wallet
    console.log("\nFetching premium data (requires payment)...");
    try {
      const premiumResponse = await client.get(
        "http://localhost:3000/premium-data"
      );
      console.log("Premium data:", premiumResponse.data);
    } catch (error) {
      console.log(
        "Payment failed (expected without funds):",
        (error as Error).message
      );
    }
  } finally {
    await client.close();
  }
}

// Example 2: Using X402Client (Manual Payment Control)
async function exampleManualClient() {
  console.log(
    "\n=== Example 2: Manual Client (Explicit Payment Control) ===\n"
  );

  const keypair = Keypair.generate();
  console.log("Wallet Address:", keypair.publicKey.toString());

  const client = new X402Client(keypair, "https://api.devnet.solana.com");

  try {
    // Initial request
    console.log("Making initial request to premium endpoint...");
    let response = await client.get("http://localhost:3000/premium-data");

    // Check if payment is required
    if (client.paymentRequired(response)) {
      console.log("Payment required!");

      // Parse payment request
      const paymentRequest = client.parsePaymentRequest(response);
      console.log("Payment details:");
      console.log("  Amount:", paymentRequest.maxAmountRequired, "USDC");
      console.log("  Recipient:", paymentRequest.paymentAddress);
      console.log("  Expires:", paymentRequest.expiresAt);

      try {
        // Create payment
        console.log("\nCreating payment...");
        const authorization = await client.createPayment(paymentRequest);
        console.log("Payment created:", authorization.transactionHash);

        // Retry with payment
        console.log("Retrying request with payment...");
        response = await client.get("http://localhost:3000/premium-data", {
          payment: authorization,
        });
        console.log("Success! Response:", response.data);
      } catch (error) {
        console.log(
          "Payment failed (expected without funds):",
          (error as Error).message
        );
      }
    } else {
      console.log("No payment required");
      console.log("Response:", response.data);
    }
  } finally {
    await client.close();
  }
}

// Example 3: Accessing Different Endpoints
async function exampleMultipleEndpoints() {
  console.log("\n=== Example 3: Multiple Endpoints ===\n");

  const keypair = Keypair.generate();
  const client = new X402AutoClient(keypair, "https://api.devnet.solana.com");

  try {
    // Root endpoint
    const root = await client.get("http://localhost:3000/");
    console.log("API Info:", root.data);

    // Free endpoint
    const free = await client.get("http://localhost:3000/free-data");
    console.log("\nFree Data:", free.data);

    // Tiered endpoint
    try {
      const tiered = await client.get(
        "http://localhost:3000/tiered-data/premium"
      );
      console.log("\nTiered Data:", tiered.data);
    } catch (error) {
      console.log(
        "\nTiered endpoint (payment required):",
        (error as Error).message
      );
    }
  } finally {
    await client.close();
  }
}

// Main execution
async function main() {
  console.log("X402 Client Examples");
  console.log("=".repeat(60));
  console.log(
    "Make sure the Express server is running on http://localhost:3000"
  );
  console.log("=".repeat(60));

  try {
    await exampleAutoClient();
    await exampleManualClient();
    await exampleMultipleEndpoints();

    console.log("\n" + "=".repeat(60));
    console.log("Examples completed!");
    console.log("=".repeat(60) + "\n");
  } catch (error) {
    console.error("Error:", error);
  }
}

// Run if executed directly
if (require.main === module) {
  main();
}

File: examples/go/nethttp-server/client_example.go
package main

import (
	"context"
	"fmt"
	"io"
	"log"
	"os"

	"github.com/gagliardetto/solana-go"
	"github.com/mr-tron/base58"
	"github.com/openlibx402/go/openlibx402-client"
)

// This example demonstrates how to use the X402 client to access paid APIs.
//
// To run this example:
//  1. Set your Solana wallet private key: export X402_PRIVATE_KEY="your-base58-private-key"
//  2. Make sure the server is running: go run main.go
//  3. Run this client: go run client_example.go
func runClientExample() {
	// Load wallet keypair from environment
	privateKeyStr := os.Getenv("X402_PRIVATE_KEY")
	if privateKeyStr == "" {
		log.Fatal("X402_PRIVATE_KEY environment variable not set")
	}

	privateKeyBytes := base58.Decode(privateKeyStr)
	walletKeypair := solana.PrivateKey(privateKeyBytes)

	// Create X402 auto client (automatically handles payments)
	client := client.NewAutoClient(walletKeypair, "", &client.AutoClientOptions{
		MaxPaymentAmount: "10.0", // Safety limit
		AutoRetry:        true,
		AllowLocal:       true, // Enable for local development
	})
	defer client.Close()

	ctx := context.Background()

	// Example 1: Access free endpoint
	log.Println("\n📡 Accessing free endpoint...")
	resp, err := client.Get(ctx, "http://localhost:8080/api/free-data")
	if err != nil {
		log.Printf("❌ Error: %v", err)
	} else {
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		log.Printf("✅ Response: %s", string(body))
	}

	// Example 2: Access premium endpoint (automatically pays)
	log.Println("\n📡 Accessing premium endpoint...")
	resp, err = client.Get(ctx, "http://localhost:8080/api/premium-data")
	if err != nil {
		log.Printf("❌ Error: %v", err)
	} else {
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		log.Printf("✅ Response: %s", string(body))
	}

	// Example 3: Access expensive endpoint
	log.Println("\n📡 Accessing expensive endpoint...")
	resp, err = client.Get(ctx, "http://localhost:8080/api/expensive-data")
	if err != nil {
		log.Printf("❌ Error: %v", err)
	} else {
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		log.Printf("✅ Response: %s", string(body))
	}

	// Example 4: POST to process endpoint
	log.Println("\n📡 Posting to process endpoint...")
	requestBody := []byte(`{"data": "sample data to process"}`)
	resp, err = client.Post(ctx, "http://localhost:8080/api/process", requestBody)
	if err != nil {
		log.Printf("❌ Error: %v", err)
	} else {
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		log.Printf("✅ Response: %s", string(body))
	}
}

// Example using explicit client (manual payment control)
func runExplicitClientExample() {
	privateKeyStr := os.Getenv("X402_PRIVATE_KEY")
	if privateKeyStr == "" {
		log.Fatal("X402_PRIVATE_KEY environment variable not set")
	}

	privateKeyBytes := base58.Decode(privateKeyStr)
	walletKeypair := solana.PrivateKey(privateKeyBytes)

	// Create explicit client
	client := client.NewX402Client(walletKeypair, "", nil, true)
	defer client.Close()

	ctx := context.Background()

	log.Println("\n📡 Accessing premium endpoint with explicit client...")

	// Initial request
	resp, err := client.Get(ctx, "http://localhost:8080/api/premium-data", nil)
	if err != nil {
		log.Fatal(err)
	}

	// Check if payment required
	if client.PaymentRequired(resp) {
		log.Println("💰 Payment required, processing payment...")

		// Parse payment request
		paymentReq, err := client.ParsePaymentRequest(resp)
		if err != nil {
			log.Fatal(err)
		}

		log.Printf("💵 Amount required: %s USDC", paymentReq.MaxAmountRequired)

		// Create and broadcast payment
		auth, err := client.CreatePayment(ctx, paymentReq, "")
		if err != nil {
			log.Fatal(err)
		}

		log.Printf("✅ Payment broadcasted: %s", auth.TransactionHash)

		// Retry with payment authorization
		resp, err = client.Get(ctx, "http://localhost:8080/api/premium-data", auth)
		if err != nil {
			log.Fatal(err)
		}
	}

	// Read response
	body, _ := io.ReadAll(resp.Body)
	resp.Body.Close()
	log.Printf("✅ Response: %s", string(body))
}

// Uncomment one of these in your main function to run the examples
func main() {
	fmt.Println("X402 Client Examples")
	fmt.Println("====================")

	// Run auto client example
	runClientExample()

	// Or run explicit client example
	// runExplicitClientExample()
}
